pkg/segment/test/segment_test.go

```diff
@@ -140,6 +258,145 @@ func TestRemoveRandom(t *testing.T) {
 	}
 }
 
+func TestAddRandomRemoveRandomHalfWithMerge(t *testing.T) {
+	var s Set
+	order := randIntervalPermutation(testSize * 2)
+	order = order[:testSize]
+	for i, j := range order {
+		if !s.Add(Range{j, j + 10}, j+valueOffset) {
+			t.Errorf("Iteration %d: failed to insert segment with key %d", i, j)
+			break
+		}
+		if err := s.checkMaxGap(); err != nil {
+			t.Errorf("When inserting %d: %v", j, err)
+			break
+		}
+	}
+	shuffle(order)
+	var nrRemovals int
+	for _, j := range order {
+		seg := s.FindSegment(j)
+		if !seg.Ok() {
+			continue
+		}
+		temprange := seg.Range()
+		s.Remove(seg)
+		nrRemovals++
+		if err := s.checkMaxGap(); err != nil {
+			t.Errorf("When removing %v: %v", temprange, err)
+			break
+		}
+	}
+	if t.Failed() {
+		t.Logf("Removal order: %v", order[:nrRemovals])
+		t.Logf("Set contents:\n%v", &s)
+		t.FailNow()
+	}
+}
+
+// Equal tells whether a and b contain the same elements.
+// A nil argument is equivalent to an empty slice.
+func Equal(a, b []int) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func TestNextLargeEnoughGap(t *testing.T) {
+	var s Set
+	order := randIntervalPermutation(testSize * 2)
+	order = order[:testSize]
+	rand.Seed(time.Now().UnixNano())
+	for i, j := range order {
+		if !s.Add(Range{j, j + rand.Intn(9) + 1}, j+valueOffset) {
+			t.Errorf("Iteration %d: failed to insert segment with key %d", i, j)
+			break
+		}
+	}
+	shuffle(order)
+	order = order[:testSize/2]
+	for _, j := range order {
+		seg := s.FindSegment(j)
+		if !seg.Ok() {
+			continue
+		}
+		s.Remove(seg)
+	}
+	minSize := 7
+	var gaparr1 []int
+	for gap := s.LowerBoundGap(0); gap.Ok(); gap = gap.NextLargeEnoughGap(minSize) {
+		if gap.Range().Length() < minSize {
+			t.Errorf("NextLargeEnoughGap wrong, gap length %d, wanted %d", gap.Range().Length(), minSize)
+		} else {
+			gaparr1 = append(gaparr1, gap.Range().Start)
+		}
+	}
+	var gaparr2 []int
+	for gap := s.LowerBoundGap(0); gap.Ok(); gap = gap.NextGap() {
+		if gap.Range().Length() >= minSize {
+			gaparr2 = append(gaparr2, gap.Range().Start)
+		}
+	}
+	if !Equal(gaparr2, gaparr1) {
+		t.Errorf("Search result not correct, got: %v, wanted: %v", gaparr1, gaparr2)
+	}
+	if t.Failed() {
+		t.Logf("Set contents:\n%v", &s)
+		t.FailNow()
+	}
+}
+
+func TestPrevLargeEnoughGap(t *testing.T) {
+	var s Set
+	order := randIntervalPermutation(testSize * 2)
+	order = order[:testSize]
+	rand.Seed(time.Now().UnixNano())
+	for i, j := range order {
+		if !s.Add(Range{j, j + rand.Intn(9) + 1}, j+valueOffset) {
+			t.Errorf("Iteration %d: failed to insert segment with key %d", i, j)
+			break
+		}
+	}
+	shuffle(order)
+	order = order[:testSize/2]
+	for _, j := range order {
+		seg := s.FindSegment(j)
+		if !seg.Ok() {
+			continue
+		}
+		s.Remove(seg)
+	}
+	minSize := 7
+	var gaparr1 []int
```

Comment:
Style: use camelCase for variable names (here and elsewhere). E.g. this would be gapArr1.

It would also be preferrable to use a descriptive named rather than "1". In this case, e.g. gapArrLargeEnough and gapArrAll would be much more descriptive and helpful.